### DISCUSSION TITLE
Deno behave differently after npm install on npm workspace

### DISCUSSION THREAD

**Discussion by bentinata** [denoland/deno#27043] [Q&A]
I would like to add deno inside a react typescript monorepo. The directory roughly looks like this:

```
package.json
apps/
  foo/
  bar/
  deno-react/
```

With this as root `package.json` contents:
```json
{
  "workspaces": [
    "apps/*"
  ]
}
```

This is `apps/deno-react/deno.json` contents, generated using `deno init --npm vite-extra` ([repository](https://github.com/bluwy/create-vite-extra/tree/master/template-deno-react-ts)):
```json
{
  "nodeModulesDir": "none",
  "tasks": {
    "dev": "deno run -A --node-modules-dir npm:vite",
    "build": "deno run -A --node-modules-dir npm:vite build",
    "preview": "deno run -A --node-modules-dir npm:vite preview",
    "serve": "deno run --allow-net --allow-read jsr:@std/http@1/file-server dist/"
  },
  "compilerOptions": {
    "lib": ["ES2020", "DOM", "DOM.Iterable"],
    "jsx": "react-jsx",
    "jsxImportSource": "react",
    "jsxImportSourceTypes": "@types/react"
  },
  "imports": {
    "@deno/vite-plugin": "npm:@deno/vite-plugin@^1.0.0",
    "@types/react": "npm:@types/react@^18.3.11",
    "@types/react-dom": "npm:@types/react-dom@^18.3.1",
    "@vitejs/plugin-react": "npm:@vitejs/plugin-react@^4.3.3",
    "react": "npm:react@^18.3.1",
    "react-dom": "npm:react-dom@^18.3.1",
    "vite": "npm:vite@^5.4.9"
  }
}
```

Without doing `npm install` in the root project directory, `deno task dev` runs just fine. But after doing `npm install`, I got this error:

```
Task dev deno run -A --node-modules-dir npm:vite
error: Uncaught (in promise) ReferenceError: require is not defined
const os = require('os');
           ^
    at file:///path-to-root-project/node_modules/supports-color/index.js:2:12
    at loadESMFromCJS (node:module:777:21)
    at Module._compile (node:module:722:12)
    at loadMaybeCjs (node:module:770:10)
    at Object.Module._extensions..js (node:module:755:12)
    at Module.load (node:module:662:32)
    at Function.Module._load (node:module:534:12)
    at Module.require (node:module:681:19)
    at require (node:module:812:16)
    at file:///path-to-root-project/apps/deno-react/node_modules/.deno/vite@5.4.10/node_modules/vite/dist/node/chunks/dep-BWSbWtLw.js:16240:26

    info: Deno supports CommonJS modules in .cjs files, or when the closest
          package.json has a "type": "commonjs" option.
    hint: Rewrite this module to ESM,
          or change the file extension to .cjs,
          or add package.json next to the
```

It'd work again after removing `node_modules` directory on root project. I've also played around `nodeModulesDir` with no success.

**Comment by nathanwhit** [comment]
What do your `foo` and `bar` package.json files look like? With what you've described I can't seem to reproduce the issue, so I think there's some missing information 

Specifically, what dependencies do you have listed in `foo` and `bar` (and do you have any dependencies in the root `package.json`)?

Have you made any changes to the `deno-react` package? For instance added dependencies or modified the config

**Comment by bentinata** [comment]
@nathanwhit 
I think it's a combination of `storybook@7` and `deno --node-modules-dir`. But it's interesting because `storybook` does not use `vite`. And more interesting that upgrading to `storybook@8` remove the issue.
I have created a minimal repositories with observation on readme here:
https://github.com/bentinata/deno-monorepo-debug
